### PR TITLE
tor source for client-agent communication

### PR DESCRIPTION
**Pull request by furkhat**

### CODE REVIEW

File: dappctrl-dev.config.json [hunk 1]
```diff
@@ -224,24 +241,13 @@
...
-    },
-    "ReportLog": {
-        "Level": "info",
-        "StackLevel": "error"
+    "SOMCServer": {
+        "Addr": "localhost:5555",
+        "TLS": null
     },
+    "StaticPassword": "",
+    "TorHostname": "plyzfy3qicjjvmeg.onion",
```
Comment (drew2a) [inline dappctrl-dev.config.json]: A Tor host name should be related to a particular Agent instance.
Why did you place this variable in config?

Comment (drew2a) [reply dappctrl-dev.config.json]: Ok, i got it.
It is just PoC.

File: agent/tor-somc/client.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package somc
```
Comment (ababo) [inline agent/tor-somc/client.go]: What about renaming the package with more idiomatic name, say `torsomc`?

File: agent/tor-somc/client.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// GetOffering gets offering message through tor net.
+func GetOffering(hostname, hash string, sock uint) (string, error) {
+	payload := fmt.Sprintf(`{"method": "api_offering",
+		"params": ["%s"], "id": 67}`, hash)
```
Comment (ababo) [inline agent/tor-somc/client.go]: Magic numbers?

File: agent/tor-somc/endpoint_test.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	fxt := data.NewTestFixture(t, db)
+	defer fxt.Close()
+
+	ch := fxt.Channel
+	keyBytes, err := (data.ChannelKey(
+		ch.Client, ch.Agent, ch.Block, fxt.Offering.Hash))
+	util.TestExpectResult(t, "ChannelKey", nil, err)
+	channelKey := data.FromBytes(keyBytes)
+
+	// TODO: store channel key in table for fast lookup?
```
Comment (sofabeat) [inline agent/tor-somc/endpoint_test.go]: Yes. It is good idea.

File: data/migration/scripts/00001_schema_up.sql [hunk 5]
```diff
@@ -226,7 +226,11 @@ CREATE TABLE offerings (
 
     additional_params json, -- all additional parameters stored as JSON -- todo: [suggestion] use jsonb to query for parameters
 
-    auto_pop_up bool -- if is true, then allow the offering to automatically pop up.
+    auto_pop_up bool, -- if is true, then allow the offering to automatically pop up
+
+    source_type smallint NOT NULL DEFAULT 0, -- defines messages communication method with agent
```
Comment (sofabeat) [inline data/migration/scripts/00001_schema_up.sql]: maybe `somc_type`? I know it comes from me, but it sound to generic to understand what is it.

File: data/migration/scripts/00001_schema_up.sql [hunk 6]
```diff
@@ -226,7 +226,11 @@ CREATE TABLE offerings (
...
+
+    source text NOT NULL, -- additional data to support defined source_type.
```
Comment (sofabeat) [inline data/migration/scripts/00001_schema_up.sql]: maybe `somc_data`? I know it comes from me, but it sound to generic to understand what is it.

File: dappctrl-dev.config.json [hunk 7]
```diff
@@ -41,7 +41,7 @@
         "CheckTimeout": 10,
         "Contract": {
             "PTCAddrHex": "0x0d825eb81b996c67a55f7da350b6e73bab3cb0ec",
-            "PSCAddrHex": "0x67c3ce713b1c40b2364360d37a15eddbc5aa9420",
+            "PSCAddrHex": "0x218d767e906d14e84683804f0a1f35a8ffdcb800",
```
Comment (dzeckelev) [inline dappctrl-dev.config.json]: New contract address `0x7ee4f7c659c1c58050d70af278e8901fc315a11e`.

File: eth/constant.go [hunk 8]
```diff
@@ -12,9 +12,9 @@ var (
 	ServiceChannelCreated            = common.HexToHash("78ef868cbadb8c73fc1170b8a09d99704c4cdf400c9c94534811b0ae60b8913a")
 	ServiceChannelToppedUp           = common.HexToHash("a3b2cd532a9050531ecc674928d7704894707ede1a436bfbee86b96b83f2a5ce")
 	ServiceChannelCloseRequested     = common.HexToHash("b40564b1d36572b2942ad7cfc5a5a967f3ef08c82163a910dee760c5b629a32e")
-	ServiceOfferingCreated           = common.HexToHash("32c1913dfde418197923027c2f2260f19903a2e86a93ed83c4689ac91a96bafd")
+	ServiceOfferingCreated           = common.HexToHash("98fd7525022191e0cb78e4245bba1fd0f2be10e6a7bf8d640151ca9d005ae73b")
```
Comment (dzeckelev) [inline eth/constant.go]: ServiceOfferingCreated           = common.HexToHash("b275d5e4029fb6389cc956ff4b5cb6ca1daa8f79b7032514dfc1c0ba99d9bd92")

File: eth/constant.go [hunk 9]
```diff
@@ -12,9 +12,9 @@ var (
...
 	ServiceOfferingDeleted           = common.HexToHash("c3013cd9dd5c33b95a9cc1bc076481c9a6a1970be6d7f1ed33adafad6e57d3d6")
-	ServiceOfferingPopedUp           = common.HexToHash("c37352067a3ca1eafcf2dc5ba537fc473509c4e4aaca729cb1dab7053ec1ffbf")
+	ServiceOfferingPopedUp           = common.HexToHash("3db9eb24898d4657e5f8f72b8584750dba18dcdc974e8cf0fa9507dc44e2e174")
```
Comment (dzeckelev) [inline eth/constant.go]: ServiceOfferingPopedUp           = common.HexToHash("12d17decebc969bcbfa8d54560cf882c7873b599dfe467aa80db3096a308edca")

File: proc/worker/event.go [hunk 10]
```diff
@@ -59,6 +64,16 @@ func init() {
 		panic(err)
 	}
 
+	abiUint8, err := abi.NewType("uint8")
+	if err != nil {
+		panic(err)
+	}
+
+	abiBytes, err := abi.NewType("bytes")
```
Comment (dzeckelev) [inline proc/worker/event.go]:
```	abiString, err := abi.NewType("string")
	if err != nil {
		panic(err)
	}
```

File: proc/worker/event.go [hunk 11]
```diff
@@ -78,6 +93,21 @@ func init() {
 		{
 			Type: abiUint16,
 		},
+		{
+			Type: abiUint8,
+		},
+		{
+			Type: abiBytes,
```
Comment (dzeckelev) [inline proc/worker/event.go]: Change to string

File: proc/worker/event.go [hunk 12]
```diff
@@ -78,6 +93,21 @@ func init() {
...
+		},
+	}
+
+	logOfferingPopUpDataArguments = abi.Arguments{
+		{
+			Type: abiUint8,
+		},
+		{
+			Type: abiBytes,
```
Comment (dzeckelev) [inline proc/worker/event.go]: Change to string

File: proc/worker/event.go [hunk 13]
```diff
@@ -174,6 +204,16 @@ func extractLogOfferingCreated(logger log.Logger,
 		return nil, ErrParseJobData
 	}
 
+	somcType, ok := dataUnpacked[1].(uint8)
+	if !ok {
+		return nil, ErrParseJobData
+	}
+
+	somcData, ok := dataUnpacked[2].([]byte)
```
Comment (dzeckelev) [inline proc/worker/event.go]: string

File: proc/worker/event.go [hunk 14]
```diff
@@ -30,17 +30,22 @@ type logOfferingCreatedInput struct {
 	offeringHash  common.Hash
 	minDeposit    *big.Int
 	currentSupply uint16
+	somcType      uint8
+	somcData        []byte
```
Comment (dzeckelev) [inline proc/worker/event.go]: Change `[]byte` to string. Update and in other places.

**Pull request merged**